dace/libraries/standard/nodes/reduce.py

```diff
@@ -1075,6 +1077,386 @@ def expansion(node: 'Reduce', state: SDFGState, sdfg: SDFG, partial_width=16):
         return nsdfg
 
 
+@dace.library.expansion
+class ExpandReduceGPUAuto(pm.ExpandTransformation):
+    """
+        GPU implementation of the reduce node. This expansion aims to map the reduction inputs to an optimal GPU schedule.
+    """
+    environments = [CUDA]
+
+    @staticmethod
+    def expansion(node: 'Reduce', state: SDFGState, sdfg: SDFG):
+        '''
+        :param node: the node to expand
+        :param state: the state in which the node is in
+        :param sdfg: the SDFG in which the node is in
+        '''
+        node.validate(sdfg, state)
+        inedge: graph.MultiConnectorEdge = state.in_edges(node)[0]
+        outedge: graph.MultiConnectorEdge = state.out_edges(node)[0]
+        insubset = dcpy(inedge.data.subset)
+        isqdim = insubset.squeeze()
+        outsubset = dcpy(outedge.data.subset)
+        osqdim = outsubset.squeeze()
+        # input_dims = len(insubset)
+        # output_dims = len(outsubset)
+        raw_input_data = sdfg.arrays[inedge.data.data]
+        raw_output_data = sdfg.arrays[outedge.data.data]
+
+        in_type = raw_input_data.dtype
+
+        if raw_input_data.storage != dtypes.StorageType.GPU_Global:
```

Comment:
Could you elaborate on that?

As far as I understand, we need to check if our whole reduce node is inside a bigger GPU kernel already, right?
And why is this bad? Because we cannot have a GPU map inside another GPU map in the SDFG?

And why does Pure not have the same issues?

Reply:
> Because we cannot have a GPU map inside another GPU map in the SDFG?

yes

> And why does Pure not have the same issues?

Pure expands to maps with `Default` schedule, these get inferred correctly later.

Reply:
Ah yes, thank you. Makes sense

Reply:
Ah yes, makes sense. Thanks